src/chains.rs

```diff
@@ -669,20 +689,228 @@ impl<'a> ChainFormatterShared<'a> {
 
         Some(result)
     }
+
+    fn has_long_item(&self) -> bool {
+        self.has_long_parent || self.has_long_inner_item || self.has_long_tail
+    }
 }
 
 // Formats a chain using block indent.
 struct ChainFormatterBlock<'a> {
     shared: ChainFormatterShared<'a>,
     root_ends_with_block: bool,
+    parent_body_forced_indent: bool,
 }
 
 impl<'a> ChainFormatterBlock<'a> {
     fn new(chain: &'a Chain) -> ChainFormatterBlock<'a> {
         ChainFormatterBlock {
             shared: ChainFormatterShared::new(chain),
             root_ends_with_block: false,
+            parent_body_forced_indent: false,
+        }
+    }
+}
+
+struct FormattedChainItem {
+    rewrite: String,
+    exceeds_max_width: bool,
+    forced_indent: bool,
+}
+
+fn format_chain_item(
+    item: &ChainItem,
+    context: &RewriteContext<'_>,
+    shape: Shape,
+) -> FormattedChainItem {
+    fn is_long(rewrite: String, max_width: usize, shape: Shape) -> bool {
+        wrap_str(rewrite, max_width, shape).is_none()
+    }
+
+    fn format_chain_parent_inner(
```

Comment:
This entire function (which runs to line 832) is to support #3157 (indentation for chains with block-like parents), all the other changes in this file are to address #3863 (long chain elements)